File: carkit/navigation/carkit_amcl/carkit_amcl/odom_tf_broadcaster.py
#!/usr/bin/env python3

import rclpy
from geometry_msgs.msg import TransformStamped
from nav_msgs.msg import Odometry
from rclpy.node import Node
from tf2_ros import TransformBroadcaster


class OdomTfBroadcaster(Node):
    def __init__(self):
        super().__init__('odom_tf_broadcaster')

        self.declare_parameter('odom_topic', '/odom')
        self.declare_parameter('odom_frame', 'odom')
        self.declare_parameter('base_frame', 'base_link')
        self.declare_parameter('use_message_stamp', True)

        self.odom_topic = self.get_parameter('odom_topic').value
        self.odom_frame = self.get_parameter('odom_frame').value
        self.base_frame = self.get_parameter('base_frame').value
        self.use_message_stamp = bool(self.get_parameter('use_message_stamp').value)

        self.tf_broadcaster = TransformBroadcaster(self)
        self.subscription = self.create_subscription(
            Odometry,
            self.odom_topic,
            self.odom_callback,
            20,
        )

        self.get_logger().info(
            f'Broadcasting {self.odom_frame} -> {self.base_frame} TF from {self.odom_topic}'
        )

    def odom_callback(self, msg):
        stamped = self._make_transform(msg)
        self.tf_broadcaster.sendTransform(stamped)

    def _make_transform(self, msg):
        stamped = TransformStamped()
        if self.use_message_stamp:
            stamped.header.stamp = msg.header.stamp
        else:
            stamped.header.stamp = self.get_clock().now().to_msg()
        stamped.header.frame_id = self.odom_frame or msg.header.frame_id
        stamped.child_frame_id = self.base_frame or msg.child_frame_id
        stamped.transform.translation.x = msg.pose.pose.position.x
        stamped.transform.translation.y = msg.pose.pose.position.y
        stamped.transform.translation.z = msg.pose.pose.position.z
        stamped.transform.rotation = msg.pose.pose.orientation
        return stamped


def main(args=None):
    rclpy.init(args=args)
    node = OdomTfBroadcaster()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: carkit/navigation/carkit_amcl/src/odom_tf_broadcaster.cpp
// Copyright 2026 University of Delaware
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <memory>
#include <string>

#include "geometry_msgs/msg/transform_stamped.hpp"
#include "nav_msgs/msg/odometry.hpp"
#include "rclcpp/rclcpp.hpp"
#include "tf2_ros/transform_broadcaster.h"

// CARKit learning annotation: implements the behavior described by this file's package and module.
class OdomTfBroadcaster : public rclcpp::Node
{
public:
  OdomTfBroadcaster()
  : Node("odom_tf_broadcaster")
  {
    const auto odom_topic = declare_parameter<std::string>("odom_topic", "/odom");
    odom_frame_ = declare_parameter<std::string>("odom_frame", "odom");
    base_frame_ = declare_parameter<std::string>("base_frame", "base_link");
    use_message_stamp_ = declare_parameter("use_message_stamp", true);
    broadcaster_ = std::make_unique<tf2_ros::TransformBroadcaster>(*this);
    subscription_ = create_subscription<nav_msgs::msg::Odometry>(
      odom_topic, 20,
      [this](nav_msgs::msg::Odometry::ConstSharedPtr message) {on_odometry(*message);});
    RCLCPP_INFO(
      get_logger(), "C++ TF adapter: %s -> %s from %s", odom_frame_.c_str(),
      base_frame_.c_str(), odom_topic.c_str());
  }

private:
  /// Republish the odometry pose as the odom-to-base transform expected by Nav2.
  void on_odometry(const nav_msgs::msg::Odometry & message)
  {
    geometry_msgs::msg::TransformStamped output;
    if (use_message_stamp_) {
      output.header.stamp = message.header.stamp;
    } else {
      output.header.stamp = now();
    }
    output.header.frame_id = odom_frame_.empty() ? message.header.frame_id : odom_frame_;
    output.child_frame_id = base_frame_.empty() ? message.child_frame_id : base_frame_;
    output.transform.translation.x = message.pose.pose.position.x;
    output.transform.translation.y = message.pose.pose.position.y;
    output.transform.translation.z = message.pose.pose.position.z;
    output.transform.rotation = message.pose.pose.orientation;
    broadcaster_->sendTransform(output);
  }

  std::string odom_frame_;
  std::string base_frame_;
  bool use_message_stamp_{true};
  std::unique_ptr<tf2_ros::TransformBroadcaster> broadcaster_;
  rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr subscription_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<OdomTfBroadcaster>());
  rclcpp::shutdown();
  return 0;
}
